### PR TITLE
Major Release: Multiple Endpoints and Code Cleanup

### PR DESCRIPTION
**Pull request by brandinarsenault**
- Renamed `callApi` to `getApi` #92 
- Removed `communityIds` parameter in `getStreams` #89 
- Removed duplicate functions in `ChannelPointsApi`
- Renamed `getEntitlmentGrantsUploadURL` to `createEntitlmentGrantsUploadURL` #93 
- Added `patchApi`
- Added `putApi`
- Added ChannelPointsApi->createCustomReward
- Added ChannelPointsApi->deleteCustomReward
- Added ChannelPointsApi->updateCustomReward
- Added ChannelPointsApi->updateRedemptionStatus
- Added ModerationApi->checkAutomodStatus
- Added StreamsApi->createStreamMarker
- Added StreamsApi->modifyChannelInfo
- Added UsersApi->updateUser

### CODE REVIEW

**Comment by brandinarsenault** [comment]
I expected this to fail as its a massive update. I think there are some issues with the body parameters which I need to resolve, please feel free to review the code in the `src` if you wish, or you can wait for the tests to pass.

**Comment by brandinarsenault** [comment]
@echosa when you have a moment can you look at the failing functions? I thought it was the way I was generating the arrays in a couple of the functions, but that fix didn't even work, so I'm not sure what I've done wrong and why these are failing. Thank you.

**Comment by echosa** [comment]
I spent some time looking into this, but can't spend any more time on it now. It seems like all the failing specs have `json_encode()` calls, so I'd start there. Something isn't working right. I've done a bit of `var_dump()` debugging to compare the values being set in the `Request` in the spec and the values being set in the `Request` in `sendToApi()`, and they look the same. If that's the case, then phpspec should be able to pick up on that. 

My running theory is that the two `Request` objects are not exactly alike, causing `phpspec` to not match them. This causes the `guzzleClient` in the spec to return `null` because phpspec doesn't find a correct matching `Request` object. Hopefully that makes sense.

**Comment by brandinarsenault** [comment]
I'm making no progress currently so I'm going to walk away for a few. So far I've:
- removed the 'Accept' header in both the Specs and the `AbstractResource`
- attempted to stop encoding, this threw an invalid resource type (expects json)
- I've verified the headers *seem* to be the same in both
- I've verified the json *seems* to be the same in both
- It was noted the only tests failing are ones with `json_encode`

Additional notes that were decided:
- Required body params will be defined in the method once we get the tests resolved

**Comment by brandinarsenault** [comment]
Just wanted to add a note as this is hitting two weeks - I've been taking some pokes here and there at this but still not finding a solution, will continue to look at it and if anyone has any ideas I'd appreciate it!

**Comment by echosa** [comment]
@brandinarsenault I've investigated this some for you. I don't have a solution, but perhaps my findings will help you create one. To make things easier on myself, I picked one spec to test and debug. I chose `ModerationApiSpec::it_should_check_automod_status()`: https://github.com/brandinarsenault/twitch-api-php/blob/444204ef1f314adf792723150e8a880b52f74e93/spec/NewTwitchApi/Resources/ModerationApiSpec.php#L18

FTR, you can run just this single spec thusly:

```
$ vendor/bin/phpspec run --no-interaction spec/NewTwitchApi/Resources/ModerationApiSpec.php:19
```

The issue is that the [Request you're creating in the spec](https://github.com/brandinarsenault/twitch-api-php/blob/444204ef1f314adf792723150e8a880b52f74e93/spec/NewTwitchApi/Resources/ModerationApiSpec.php#L20) isn't matching the [Request created by the library](https://github.com/brandinarsenault/twitch-api-php/blob/42f132288515c2e1deee2e98ce4c582d3f4332ea/src/NewTwitchApi/Resources/AbstractResource.php#L64-L71). Therefore, when phpspec defines "for request X, return Response", it never matches because the library is creating similar, but not the same, request Y.

The easiest way to demonstrate this is replace the phpspec `$guzzleClient` definition with this instead:         
```
        $guzzleClient->send(Argument::any())->willReturn($response);
```

This tells phpspec to simply expect any Request at all. You will notice that spec will now pass, because we're not specifying anything specific about the Request itself, only that `$guzzleClient::send()` is called. (Which it is.) This proves that the conflicting difference is with the Request objects. So, I added some `var_dump` calls to dump the Request data that are being set. First, in the spec:

```
    function it_should_check_automod_status(Client $guzzleClient, Response $response)
    {
        $request = new Request('POST', 'moderation/enforcements/status?broadcaster_id=123', ['Authorization' => 'Bearer TEST_TOKEN', 'Accept' => 'application/json'], json_encode(['msg_id' => '321', 'msg_text' => 'test', 'user_id' => '111']));
        $guzzleClient->send(Argument::any())->willReturn($response);
        var_dump('In PhpSpec');
        var_dump($request->getMethod());
        var_dump((string) $request->getUri());
        var_dump($request->getHeaders());
        var_dump($request->getBody()->getContents());
        $this->checkAutomodStatus('TEST_TOKEN', '123', '321', 'test', '111')->shouldBeAnInstanceOf(ResponseInterface::class);
    }
```

and also in the logic code:

```
    private function sendToApi(string $httpMethod, string $uriEndpoint, string $bearer, array $queryParamsMap = [], array $bodyParams = []): ResponseInterface
    {
        if (count($bodyParams) > 0) {
// ...
        }

        var_dump('In Logic Code');
        var_dump($request->getMethod());
        var_dump((string) $request->getUri());
        var_dump($request->getHeaders());
        var_dump($request->getBody()->getContents());

        return $this->guzzleClient->send($request);
    }
```

This is the output:

```
string(10) "In PhpSpec"
string(4) "POST"
string(49) "moderation/enforcements/status?broadcaster_id=123"
array(2) {
  ["Authorization"]=>
  array(1) {
    [0]=>
    string(17) "Bearer TEST_TOKEN"
  }
  ["Accept"]=>
  array(1) {
    [0]=>
    string(16) "application/json"
  }
}
string(50) "{"msg_id":"321","msg_text":"test","user_id":"111"}"
string(13) "In Logic Code"
string(4) "POST"
string(49) "moderation/enforcements/status?broadcaster_id=123"
array(2) {
  ["Authorization"]=>
  array(1) {
    [0]=>
    string(17) "Bearer TEST_TOKEN"
  }
  ["Accept"]=>
  array(1) {
    [0]=>
    string(16) "application/json"
  }
}
string(56) "[{"msg_id":"321"},{"msg_text":"test"},{"user_id":"111"}]"
```

Comparing the values, the only difference I see is the body contents:

```
string(50) "{"msg_id":"321","msg_text":"test","user_id":"111"}"
// versus
string(56) "[{"msg_id":"321"},{"msg_text":"test"},{"user_id":"111"}]"
```

So, we can see that at least the body isn't correct. Additionally, this makes sense, since body is a newly added construct to the code, so it seems a likely candidate for issues. Doubly so, since the only specs that seem to be failing are the new non-GET ones.

This is by no means saying the body is the _only_ issue, but it certainly seems to be _an_ issue. I suggest figuring out and fixing this issue and see if that solves the problem. 

I've exhausted the time I have to spend on this right now. Hopefully this helps.

**Comment by echosa** [comment]
Additionally, please correct the code formatting of the `sprintf` calls in both the `if` and `else` blocks here: https://github.com/brandinarsenault/twitch-api-php/blob/42f132288515c2e1deee2e98ce4c582d3f4332ea/src/NewTwitchApi/Resources/AbstractResource.php#L66-L68

The way it is now is difficult to read, and makes it look like `Request()` has six arguments instead of four:

```
            $request = new Request(
                $httpMethod,
                sprintf('%s%s',
                $uriEndpoint,
                $this->generateQueryParams($queryParamsMap)),
                ['Authorization' => sprintf('Bearer %s', $bearer), 'Accept' => 'application/json'],
                json_encode($bodyParams)
            );
```

This is much more readable, and more clear that there are only four arguments being passed to `Request()`:

```
            $request = new Request(
                $httpMethod,
                sprintf(
                    '%s%s',
                    $uriEndpoint,
                    $this->generateQueryParams($queryParamsMap)
                ),
                ['Authorization' => sprintf('Bearer %s', $bearer), 'Accept' => 'application/json'],
                json_encode($bodyParams)
            );
```

**Comment by brandinarsenault** [comment]
The body contents not matching seems likely to be the issue, since all the other specs are working as expected. This puts me in the right direction - thank you. Let me get cracking at this.

**Comment by brandinarsenault** [comment]
I've resolved the issue with the body not matching, unfortunately, it hasn't resolved the test from failing but its progress...I'll push those changes and then move forward from there.

**Comment by brandinarsenault** [comment]
Hi guys - I've continued looking into this a couple of hours per week since my last comment and commits 3 weeks ago - no success on my part. The body and headers are the same as far as I can tell. I've also checked the Protocol Version and encoded the data in a dump to see if somehow during that process it got adjusted. I'm unsure where to go here, a second pair of eyes is appreciated.

**Comment by brandinarsenault** [comment]
I've decided to get this large release merged I'm temporarily pulling the new features with body params as they're what is causing the issue. They were all additions, we can do that later. However, after pushing this, there seems to be an issue with GA Actions - the tests are not running, producing this error:
`update-alternatives: error: alternative /usr/bin/php7.3 for php not registered`

A Google Search didn't yield me much info, any thoughts on how we resolve this?

**Comment by echosa** [comment]
I'm cool with removing the failing bits and updating the test suite to remove 7.2 and 7.3 in the other PR. That will allow this to not stagnate any longer (wish I'd thought of it first... ;-) ), and as long as we release this under a new major version, we shouldn't break too many user setups.

**Comment by brandinarsenault** [comment]
@echosa do you mind doing one last review before I merge this?

**Comment by brandinarsenault** [comment]
IP release notes:
```markdown
**BREAKING CHANGES**
- Removed `communityIds` parameter in `getStreams` #89 
- Renamed `getEntitlmentGrantsUploadURL` to `createEntitlmentGrantsUploadURL` #93 

**NEW FUNCTIONS**
- Added ChannelPointsApi->createCustomReward
- Added ChannelPointsApi->deleteCustomReward
- Added UsersApi->updateUser

**MAINTENANCE**
- Removed duplicate functions in `ChannelPointsApi`
- Renamed `callApi` to `getApi` #92 
- Added `patchApi` ability in library
- Added `postApi` ability in library
- Added `putApi` ability in library
- Added `deleteApi` ability in library

**VERSION NOTE**
Support for PHP 7.2 and 7.3 are no longer maintained and are not tested in our workflows. While there are no known issues at this time, we cannot guarantee this library will work on a version less than PHP 7.4 at the time of writing.
```

File: src/NewTwitchApi/HelixGuzzleClient.php
Comment (echosa) [inline src/NewTwitchApi/HelixGuzzleClient.php:11]: What is this? This is causing tests to fail, thankfully. I'm pretty sure we want to send Twitch API calls to Twitch's API, right?

Comment (brandinarsenault) [reply src/NewTwitchApi/HelixGuzzleClient.php]: This was a local change I made to test the Body Params and accidentally pushed when I used my "stage all" button, thank goodness for tests! Rolling it back now.

**Comment by brandinarsenault** [comment]
@echosa resolved the BASE_URI issue, good to finish your review.

**Review by echosa — APPROVED** [review]
Thank you for all the work!

**Pull request merged**
